### PR TITLE
Handle precision loss during sweep transaction fee distribution

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Closes: #150 
Depends on: #162 

So far, the sweep transaction fee was distributed over all deposits but the potential remainder was lost due to Solidity rounding. This change mitigates that problem by isolating the indivisible remainder and pushing it towards the last swept deposit determined according to the order of the inputs.

### CODE REVIEW

**Comment by lukasz-zimnoch** [comment]
Ready for review but do not merge until #162 lands on main.

File: solidity/contracts/bridge/Bridge.sol
Comment (tomaszslabon) [inline solidity/contracts/bridge/Bridge.sol:682]: `depositTxFeeRemainder` will be at most `depositedAmounts.length - 1`.
If we assume the number of swept deposits is no greater than `10`, then `depositTxFeeRemainder` will be at most `9`.

So if we have 10 deposits, and tx fee is 999, the cost will spread like this:
99, 99, 99, 99, 99, 99, 99, 99, 99, 108.

9 satoshis paid more is not much and we can live with this.

Alternatively we could:
1) Force the wallet to always choose a fee that is divisible by the number of deposits
2) Or spread the cost more evenly:
  99 100 100 100 100 100 100 100 100 100

 So that the remainder is split evenly among the last depositors. 
 If the remainder is 9, add 1 to the cost of the last 9 depositors.
 If the remainder is 4, add 1 to the cost of the last 4 depositors.
 If the remainder is 1, add 1 to the cost of the last depositor.

Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:682]: Yep, that's right, there are some on-chain alternatives. However, pushing the remainder to the last deposit seems to be the simplest and most gas-effective option at the moment. Currently, 9 satoshi is ~0.00354 USD so I think the additional cost for the last deposit is negligible. We can also implement the "divisible fee" behavior in the off-chain client at any moment.

Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:682]: Although I do agree with @tomaszslabon that the two options he proposed are more fair, I also agree with @lukasz-zimnoch that we should choose the least expensive option given the gas consumption. Eventually, it is the depositor who will pay in the treasury fee for the gas consumption of the sweep proof transaction so option 2 could end up as theoretically more fair but practically, depositors would receive less value. Option 1 feels too dangerous to operators, in my opinion. The off-chain signing protocol will be complex enough even without additional constraints on the fee.

File: solidity/test/bridge/Bridge.test.ts
Comment (pdyraga) [inline solidity/test/bridge/Bridge.test.ts:895]: Clear example that `depositTxMaxFee = 1000; // 1000 satoshi` in the constructor does not make sense 😬 

Not in this PR, but can we increase it to 100k? I know there is a TODO to revisit these values but 1000 is definitely not enough.

**Review by pdyraga — APPROVED** [review]
Looks good to me, leaving it in @tomaszslabon's hands.

File: solidity/test/bridge/Bridge.test.ts
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts:895]: Yeah this was just an initial value that will be revisited anyway. Also remember `depositTxMaxFee` is max transaction fee per deposit. Not for the entire sweep transaction.

File: solidity/test/bridge/Bridge.test.ts
Comment (tomaszslabon) [inline solidity/test/bridge/Bridge.test.ts:898]: Shouldn't it be 603 instead of 604?

Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.test.ts]: Yep, right. Fixed https://github.com/keep-network/tbtc-v2/pull/164/commits/2e8a2b2f9a3a033fe4ddd8f76ba966394d839b8e

**Review by tomaszslabon — APPROVED** [review]

**Pull request merged**
